File: tools/covenant/feedback_compiler.py
#!/usr/bin/env python3
"""Agent Feedback Compiler (AFC).

Compiles agent-specific lessons from correction/task history and provides
query/injection interfaces for spawn prompt enrichment.
"""

from __future__ import annotations

import argparse
import json
import os
import re
import subprocess
from typing import Iterable

DB_NAME = "cortana"
DB_BIN = "/opt/homebrew/opt/postgresql@17/bin"

AGENT_KEYWORDS: dict[str, tuple[str, ...]] = {
    "huragok": ("huragok", "implement", "build", "code", "migration", "infra", "branch", "git"),
    "researcher": ("researcher", "research", "sources", "evidence", "findings", "synthesis"),
    "librarian": ("librarian", "docs", "documentation", "readme", "spec", "runbook"),
    "oracle": ("oracle", "forecast", "risk", "strategy", "decision", "model"),
    "monitor": ("monitor", "monitoring", "alert", "anomaly", "health", "watchdog"),
}

TASK_ISSUE_PATTERNS = (
    "issue",
    "problem",
    "failed",
    "failure",
    "regression",
    "backlog",
    "bug",
    "retry",
)


def sql_escape(text: str) -> str:
    return (text or "").replace("'", "''")


def run_psql(sql: str) -> str:
    env = os.environ.copy()
    env["PATH"] = f"{DB_BIN}:{env.get('PATH', '')}"
    cmd = ["psql", DB_NAME, "-q", "-X", "-v", "ON_ERROR_STOP=1", "-t", "-A", "-c", sql]
    out = subprocess.run(cmd, text=True, capture_output=True, env=env)
    if out.returncode != 0:
        raise RuntimeError(out.stderr.strip() or out.stdout.strip() or "psql failed")
    return out.stdout.strip()


def classify_agent(text: str) -> str:
    lowered = (text or "").lower()
    hits: dict[str, int] = {}
    for role, words in AGENT_KEYWORDS.items():
        score = sum(1 for w in words if w in lowered)
        if score:
            hits[role] = score
    if not hits:
        return "all"
    return sorted(hits.items(), key=lambda kv: kv[1], reverse=True)[0][0]


def normalize_lesson(text: str) -> str:
    cleaned = re.sub(r"\s+", " ", (text or "").strip())
    return cleaned[:800]


def upsert_lesson(
    *,
    agent_role: str,
    feedback_text: str,
    confidence: float,
    source_feedback_id: int | None = None,
    source_task_id: int | None = None,
) -> None:
    lesson = normalize_lesson(feedback_text)
    if not lesson:
        return

    source_feedback_sql = "NULL" if source_feedback_id is None else str(source_feedback_id)
    source_task_sql = "NULL" if source_task_id is None else str(source_task_id)

    sql = f"""
INSERT INTO cortana_agent_feedback
  (agent_role, feedback_text, source_feedback_id, source_task_id, confidence, active, updated_at)
VALUES
  ('{sql_escape(agent_role)}', '{sql_escape(lesson)}', {source_feedback_sql}, {source_task_sql}, {confidence:.2f}, TRUE, NOW())
ON CONFLICT (agent_role, lower(feedback_text), active) WHERE active = TRUE
DO UPDATE
SET confidence = GREATEST(cortana_agent_feedback.confidence, EXCLUDED.confidence),
    source_feedback_id = COALESCE(EXCLUDED.source_feedback_id, cortana_agent_feedback.source_feedback_id),
    source_task_id = COALESCE(EXCLUDED.source_task_id, cortana_agent_feedback.source_task_id),
    updated_at = NOW();
"""
    run_psql(sql)


def _parse_rows(raw: str) -> list[list[str]]:
    rows: list[list[str]] = []
    for line in (raw or "").splitlines():
        if not line.strip():
            continue
        rows.append(line.split("|"))
    return rows


def compile_from_feedback() -> int:
    sql = """
SELECT id, feedback_type, context, lesson, applied
FROM cortana_feedback
ORDER BY id DESC
LIMIT 300;
"""
    rows = _parse_rows(run_psql(sql))
    inserted = 0

    for row in rows:
        feedback_id = int(row[0])
        feedback_type = row[1] if len(row) > 1 else ""
        context = row[2] if len(row) > 2 else ""
        lesson = row[3] if len(row) > 3 else ""
        applied = (row[4] if len(row) > 4 else "").lower() in ("t", "true", "1")

        if not lesson:
            continue

        combined = f"{feedback_type} {context} {lesson}"
        role = classify_agent(combined)
        confidence = 0.88 if applied else 0.72
        upsert_lesson(
            agent_role=role,
            feedback_text=lesson,
            confidence=confidence,
            source_feedback_id=feedback_id,
        )
        inserted += 1

    return inserted


def compile_from_tasks() -> int:
    pattern_sql = " OR ".join(
        [f"COALESCE(outcome,'') ILIKE '%{p}%'" for p in TASK_ISSUE_PATTERNS]
        + [f"COALESCE(description,'') ILIKE '%{p}%'" for p in TASK_ISSUE_PATTERNS]
    )

    sql = f"""
SELECT id, title, description, status, outcome
FROM cortana_tasks
WHERE status = 'failed'
   OR (status = 'completed' AND ({pattern_sql}))
ORDER BY id DESC
LIMIT 300;
"""

    rows = _parse_rows(run_psql(sql))
    inserted = 0

    for row in rows:
        task_id = int(row[0])
        title = row[1] if len(row) > 1 else ""
        description = row[2] if len(row) > 2 else ""
        status = row[3] if len(row) > 3 else ""
        outcome = row[4] if len(row) > 4 else ""

        combined = f"{title} {description} {outcome}"
        role = classify_agent(combined)

        if status == "failed":
            lesson = f"Avoid repeat failure pattern from task '{title}': {outcome or description or 'Investigate root cause before retry.'}"
            confidence = 0.82
        else:
            lesson = f"For task '{title}', preserve this learning from issues encountered: {outcome or description}"
            confidence = 0.76

        upsert_lesson(
            agent_role=role,
            feedback_text=lesson,
            confidence=confidence,
            source_task_id=task_id,
        )
        inserted += 1

    return inserted


def cmd_compile(_: argparse.Namespace) -> int:
    f_count = compile_from_feedback()
    t_count = compile_from_tasks()
    print(json.dumps({"compiled_from_feedback": f_count, "compiled_from_tasks": t_count, "total": f_count + t_count}))
    return 0


def query_lessons(agent_role: str, limit: int = 5) -> list[dict[str, str | int | float]]:
    role = sql_escape(agent_role.lower())
    sql = f"""
SELECT id, agent_role, feedback_text, confidence, created_at
FROM cortana_agent_feedback
WHERE active = TRUE
  AND (agent_role = '{role}' OR agent_role = 'all')
ORDER BY confidence DESC, updated_at DESC
LIMIT {int(limit)};
"""
    rows = _parse_rows(run_psql(sql))
    out: list[dict[str, str | int | float]] = []
    for row in rows:
        out.append(
            {
                "id": int(row[0]),
                "agent_role": row[1],
                "feedback_text": row[2],
                "confidence": float(row[3]),
                "created_at": row[4],
            }
        )
    return out


def cmd_query(args: argparse.Namespace) -> int:
    items = query_lessons(args.agent_role, limit=args.limit)
    print(json.dumps(items, indent=2))
    return 0


def build_injection_block(agent_role: str, limit: int = 5) -> str:
    items = query_lessons(agent_role, limit=limit)
    if not items:
        return "## Agent Feedback Lessons\n- No active lessons available for this role yet."

    lines = ["## Agent Feedback Lessons", "These are curated lessons from prior corrections and task outcomes. Apply them in this run."]
    for i, item in enumerate(items, start=1):
        lines.append(f"{i}. {item['feedback_text']} (confidence={item['confidence']:.2f})")
    return "\n".join(lines)


def cmd_inject(args: argparse.Namespace) -> int:
    print(build_injection_block(args.agent_role, limit=args.limit))
    return 0


def cmd_deactivate(args: argparse.Namespace) -> int:
    run_psql(
        f"UPDATE cortana_agent_feedback SET active = FALSE, updated_at = NOW() WHERE id = {int(args.id)};"
    )
    print(json.dumps({"deactivated": int(args.id)}))
    return 0


def cmd_stats(_: argparse.Namespace) -> int:
    sql = """
SELECT agent_role, COUNT(*)
FROM cortana_agent_feedback
WHERE active = TRUE
GROUP BY agent_role
ORDER BY COUNT(*) DESC, agent_role ASC;
"""
    rows = _parse_rows(run_psql(sql))
    payload = [{"agent_role": r[0], "active_lessons": int(r[1])} for r in rows]
    print(json.dumps(payload, indent=2))
    return 0


def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Agent Feedback Compiler (AFC)")
    sub = p.add_subparsers(dest="command", required=True)

    sp_compile = sub.add_parser("compile", help="Compile lessons from cortana_feedback and cortana_tasks")
    sp_compile.set_defaults(func=cmd_compile)

    sp_query = sub.add_parser("query", help="Query top active lessons for an agent role")
    sp_query.add_argument("agent_role", help="Agent role, e.g. huragok/researcher/librarian/oracle/monitor")
    sp_query.add_argument("--limit", type=int, default=5)
    sp_query.set_defaults(func=cmd_query)

    sp_inject = sub.add_parser("inject", help="Render spawn-ready instruction block for an agent role")
    sp_inject.add_argument("agent_role", help="Agent role")
    sp_inject.add_argument("--limit", type=int, default=5)
    sp_inject.set_defaults(func=cmd_inject)

    sp_deactivate = sub.add_parser("deactivate", help="Deactivate a lesson by id")
    sp_deactivate.add_argument("id", type=int)
    sp_deactivate.set_defaults(func=cmd_deactivate)

    sp_stats = sub.add_parser("stats", help="Show active lesson counts per agent role")
    sp_stats.set_defaults(func=cmd_stats)

    return p


def main() -> int:
    parser = build_parser()
    args = parser.parse_args()
    return int(args.func(args))


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/covenant/feedback_compiler.ts
#!/usr/bin/env npx tsx

/** Agent Feedback Compiler (AFC). */

import { runPsql, withPostgresPath } from "../lib/db.js";

const DB_NAME = "cortana";

const AGENT_KEYWORDS: Record<string, string[]> = {
  huragok: ["huragok", "implement", "build", "code", "migration", "infra", "branch", "git"],
  researcher: ["researcher", "research", "sources", "evidence", "findings", "synthesis"],
  librarian: ["librarian", "docs", "documentation", "readme", "spec", "runbook"],
  oracle: ["oracle", "forecast", "risk", "strategy", "decision", "model"],
  monitor: ["monitor", "monitoring", "alert", "anomaly", "health", "watchdog"],
};

const TASK_ISSUE_PATTERNS = [
  "issue",
  "problem",
  "failed",
  "failure",
  "regression",
  "backlog",
  "bug",
  "retry",
];

function sqlEscape(text: string): string {
  return (text || "").replace(/'/g, "''");
}

function runPsqlQuery(sql: string): string {
  const result = runPsql(sql, {
    db: DB_NAME,
    args: ["-q", "-X", "-v", "ON_ERROR_STOP=1", "-t", "-A"],
    env: withPostgresPath(process.env),
  });
  if (result.status !== 0) {
    const err = (result.stderr || result.stdout || "psql failed").toString().trim();
    throw new Error(err || "psql failed");
  }
  return (result.stdout || "").toString().trim();
}

function classifyAgent(text: string): string {
  const lowered = (text || "").toLowerCase();
  const hits: Record<string, number> = {};
  for (const [role, words] of Object.entries(AGENT_KEYWORDS)) {
    const score = words.reduce((acc, w) => acc + (lowered.includes(w) ? 1 : 0), 0);
    if (score) hits[role] = score;
  }
  if (!Object.keys(hits).length) return "all";
  return Object.entries(hits).sort((a, b) => b[1] - a[1])[0][0];
}

function normalizeLesson(text: string): string {
  return (text || "").trim().replace(/\s+/g, " ").slice(0, 800);
}

function upsertLesson(params: {
  agentRole: string;
  feedbackText: string;
  confidence: number;
  sourceFeedbackId?: number | null;
  sourceTaskId?: number | null;
}): void {
  const lesson = normalizeLesson(params.feedbackText);
  if (!lesson) return;

  const sourceFeedbackSql = params.sourceFeedbackId == null ? "NULL" : String(params.sourceFeedbackId);
  const sourceTaskSql = params.sourceTaskId == null ? "NULL" : String(params.sourceTaskId);

  const sql = `
INSERT INTO cortana_agent_feedback
  (agent_role, feedback_text, source_feedback_id, source_task_id, confidence, active, updated_at)
VALUES
  ('${sqlEscape(params.agentRole)}', '${sqlEscape(lesson)}', ${sourceFeedbackSql}, ${sourceTaskSql}, ${params.confidence.toFixed(2)}, TRUE, NOW())
ON CONFLICT (agent_role, lower(feedback_text), active) WHERE active = TRUE
DO UPDATE
SET confidence = GREATEST(cortana_agent_feedback.confidence, EXCLUDED.confidence),
    source_feedback_id = COALESCE(EXCLUDED.source_feedback_id, cortana_agent_feedback.source_feedback_id),
    source_task_id = COALESCE(EXCLUDED.source_task_id, cortana_agent_feedback.source_task_id),
    updated_at = NOW();
`;
  runPsqlQuery(sql);
}

function parseRows(raw: string): string[][] {
  const rows: string[][] = [];
  for (const line of (raw || "").split(/\r?\n/)) {
    if (!line.trim()) continue;
    rows.push(line.split("|"));
  }
  return rows;
}

function compileFromFeedback(): number {
  const sql = `
SELECT id, feedback_type, context, lesson, applied
FROM cortana_feedback
ORDER BY id DESC
LIMIT 300;
`;
  const rows = parseRows(runPsqlQuery(sql));
  let inserted = 0;

  for (const row of rows) {
    const feedbackId = Number(row[0]);
    const feedbackType = row[1] ?? "";
    const context = row[2] ?? "";
    const lesson = row[3] ?? "";
    const applied = (row[4] ?? "").toLowerCase();
    const appliedBool = applied === "t" || applied === "true" || applied === "1";

    if (!lesson) continue;

    const combined = `${feedbackType} ${context} ${lesson}`;
    const role = classifyAgent(combined);
    const confidence = appliedBool ? 0.88 : 0.72;
    upsertLesson({
      agentRole: role,
      feedbackText: lesson,
      confidence,
      sourceFeedbackId: feedbackId,
    });
    inserted += 1;
  }

  return inserted;
}

function compileFromTasks(): number {
  const patternSql = TASK_ISSUE_PATTERNS.map((p) => `COALESCE(outcome,'') ILIKE '%${p}%'`)
    .concat(TASK_ISSUE_PATTERNS.map((p) => `COALESCE(description,'') ILIKE '%${p}%'`))
    .join(" OR ");

  const sql = `
SELECT id, title, description, status, outcome
FROM cortana_tasks
WHERE status = 'failed'
   OR (status = 'completed' AND (${patternSql}))
ORDER BY id DESC
LIMIT 300;
`;

  const rows = parseRows(runPsqlQuery(sql));
  let inserted = 0;

  for (const row of rows) {
    const taskId = Number(row[0]);
    const title = row[1] ?? "";
    const description = row[2] ?? "";
    const status = row[3] ?? "";
    const outcome = row[4] ?? "";

    const combined = `${title} ${description} ${outcome}`;
    const role = classifyAgent(combined);

    let lesson: string;
    let confidence: number;
    if (status === "failed") {
      lesson = `Avoid repeat failure pattern from task '${title}': ${outcome || description || "Investigate root cause before retry."}`;
      confidence = 0.82;
    } else {
      lesson = `For task '${title}', preserve this learning from issues encountered: ${outcome || description}`;
      confidence = 0.76;
    }

    upsertLesson({
      agentRole: role,
      feedbackText: lesson,
      confidence,
      sourceTaskId: taskId,
    });
    inserted += 1;
  }

  return inserted;
}

function cmdCompile(): number {
  const fCount = compileFromFeedback();
  const tCount = compileFromTasks();
  console.log(
    JSON.stringify({ compiled_from_feedback: fCount, compiled_from_tasks: tCount, total: fCount + tCount })
  );
  return 0;
}

function queryLessons(agentRole: string, limit = 5): Array<Record<string, string | number>> {
  const role = sqlEscape(agentRole.toLowerCase());
  const sql = `
SELECT id, agent_role, feedback_text, confidence, created_at
FROM cortana_agent_feedback
WHERE active = TRUE
  AND (agent_role = '${role}' OR agent_role = 'all')
ORDER BY confidence DESC, updated_at DESC
LIMIT ${Math.trunc(limit)};
`;
  const rows = parseRows(runPsqlQuery(sql));
  return rows.map((row) => ({
    id: Number(row[0]),
    agent_role: row[1],
    feedback_text: row[2],
    confidence: Number(row[3]),
    created_at: row[4],
  }));
}

function cmdQuery(agentRole: string, limit: number): number {
  const items = queryLessons(agentRole, limit);
  console.log(JSON.stringify(items, null, 2));
  return 0;
}

function buildInjectionBlock(agentRole: string, limit = 5): string {
  const items = queryLessons(agentRole, limit);
  if (!items.length) {
    return "## Agent Feedback Lessons\n- No active lessons available for this role yet.";
  }

  const lines = [
    "## Agent Feedback Lessons",
    "These are curated lessons from prior corrections and task outcomes. Apply them in this run.",
  ];
  items.forEach((item, idx) => {
    lines.push(`${idx + 1}. ${item.feedback_text} (confidence=${Number(item.confidence).toFixed(2)})`);
  });
  return lines.join("\n");
}

function cmdInject(agentRole: string, limit: number): number {
  compileFromFeedback();
  compileFromTasks();
  console.log(buildInjectionBlock(agentRole, limit));
  return 0;
}

function cmdDeactivate(id: number): number {
  runPsqlQuery(
    `UPDATE cortana_agent_feedback SET active = FALSE, updated_at = NOW() WHERE id = ${Math.trunc(id)};`
  );
  console.log(JSON.stringify({ deactivated: Math.trunc(id) }));
  return 0;
}

function cmdStats(): number {
  const sql = `
SELECT agent_role, COUNT(*)
FROM cortana_agent_feedback
WHERE active = TRUE
GROUP BY agent_role
ORDER BY COUNT(*) DESC, agent_role ASC;
`;
  const rows = parseRows(runPsqlQuery(sql));
  const payload = rows.map((r) => ({ agent_role: r[0], active_lessons: Number(r[1]) }));
  console.log(JSON.stringify(payload, null, 2));
  return 0;
}

function usageError(): never {
  console.error("usage: feedback_compiler.ts {compile|query|inject|deactivate|stats} [args]");
  process.exit(2);
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const command = args.shift();
  if (!command) usageError();

  if (command === "compile") {
    process.exit(cmdCompile());
  }

  if (command === "query" || command === "inject") {
    const agentRole = args[0];
    if (!agentRole) usageError();
    const limitIdx = args.indexOf("--limit");
    let normalizedLimit = 5;
    if (limitIdx >= 0) {
      const raw = args[limitIdx + 1];
      if (!raw) usageError();
      const parsed = Number.parseInt(raw, 10);
      if (Number.isNaN(parsed)) usageError();
      normalizedLimit = parsed;
    }
    process.exit(command === "query" ? cmdQuery(agentRole, normalizedLimit) : cmdInject(agentRole, normalizedLimit));
  }

  if (command === "deactivate") {
    const id = args[0];
    if (!id) usageError();
    const parsed = Number.parseInt(id, 10);
    if (Number.isNaN(parsed)) usageError();
    process.exit(cmdDeactivate(parsed));
  }

  if (command === "stats") {
    process.exit(cmdStats());
  }

  usageError();
}

main().catch((err) => {
  console.error(err instanceof Error ? err.stack || err.message : String(err));
  process.exit(1);
});
